server/src/main/java/io/crate/types/NumericStorage.java

```diff
@@ -0,0 +1,193 @@
+ * to support using {@link PointRangeQuery}
+ * </li>
+ * </p>
+ *
+ **/
+public final class NumericStorage extends StorageSupport<BigDecimal> {
+
+    public static final int COMPACT_PRECISION = 18;
+
+    public NumericStorage() {
+        super(true, true, new NumericEqQuery());
+    }
+
+    @Override
+    public ValueIndexer<? super BigDecimal> valueIndexer(RelationName table,
+                                                         Reference ref,
+                                                         Function<ColumnIdent, Reference> getRef) {
+        DataType<?> type = ArrayType.unnest(ref.valueType());
+        assert type instanceof NumericType
+            : "ValueIndexer on NumericStorage can only be used for numeric types";
+        NumericType numericType = (NumericType) type;
+        Integer precision = numericType.numericPrecision();
+        if (precision == null || numericType.scale() == null) {
+            throw new UnsupportedOperationException(
+                "NUMERIC type requires precision and scale to support storage");
+        } else if (precision <= COMPACT_PRECISION) {
+            return new CompactNumericIndexer(ref);
+        } else {
+            return new FixedNumericIndexer(ref, precision);
+        }
+    }
+
+    private abstract static class BaseNumericIndexer implements ValueIndexer<BigDecimal> {
+
+        protected final Reference ref;
+        protected final String name;
+
+        protected BaseNumericIndexer(Reference ref) {
+            this.ref = ref;
+            this.name = ref.storageIdent();
+        }
+
+        @Override
+        public String storageIdentLeafName() {
+            return ref.storageIdentLeafName();
+        }
+    }
+
+    private static class CompactNumericIndexer extends BaseNumericIndexer {
+
+        private CompactNumericIndexer(Reference ref) {
+            super(ref);
+        }
+
+        @Override
+        public void indexValue(@NotNull BigDecimal value, IndexDocumentBuilder docBuilder) throws IOException {
+            BigInteger unscaled = value.unscaledValue();
+            long longValue = unscaled.longValueExact();
+
+            if (this.ref.indexType() != IndexType.NONE) {
+                docBuilder.addField(new LongPoint(name, longValue));
+            }
+            if (ref.hasDocValues()) {
+                docBuilder.addField(new SortedNumericDocValuesField(name, longValue));
+            } else {
+                docBuilder.addField(new Field(
+                    DocSysColumns.FieldNames.NAME,
+                    name,
+                    DocSysColumns.FieldNames.FIELD_TYPE));
+            }
+
+            docBuilder.translogWriter().writeValue(value);
+        }
+    }
+
+    private static class FixedNumericIndexer extends BaseNumericIndexer {
+
+        private final FieldType fieldType;
+
+        private FixedNumericIndexer(Reference ref, int precision) {
+            super(ref);
+            this.fieldType = new FieldType();
```

Comment:
I think we'll have to set dimensions here too to make range queries work, but I'll extend that in the next PR.

Reply:
You can use a `BinaryPoint` which doesn't need a `FieldType` to be defined.  It also has static methods for generating exact and range queries.

Reply:
I'm not sure if that will work out. As far as I understood `BinaryPoint` infers the `bytesPerDim` from the provided `byte[]` - given that the `BigInteger` array is variable length we'd end up with different dimension/fields per document. That would cause problems, no?

To address that `BigIntegerPoint` scales the dimension to a fixed 16 byte (using `NumericUtils.bigIntToSortableBytes`)
Given that we may have more than 16 byte my plan was to introduce some logic to convert precision to max-byte-length and then re-use `bigIntToSortableBytes`